File: 290.Word-Pattern.go
// https://leetcode.com/problems/increasing-triplet-subsequence/description/
//
// algorithms
// Medium (34.34%)
// Total Accepted:    126,215
// Total Submissions: 367,525
// beats 100.0% of golang submissions

package leetcode

import "strings"

func wordPattern(pattern string, str string) bool {
	strMap := make(map[string]byte)
	patternMap := make(map[byte]string)

	strArr := strings.Fields(str)

	if len(pattern) != len(strArr) {
		return false
	}

	for idx, s := range strArr {
		_, strExist := strMap[s]
		_, patternExist := patternMap[pattern[idx]]
		if !strExist && !patternExist {
			strMap[s] = pattern[idx]
			patternMap[pattern[idx]] = s
		} else if patternExist && strExist {
			if strMap[s] != pattern[idx] || patternMap[pattern[idx]] != s {
				return false
			}
		} else {
			return false
		}
	}

	return true
}

File: 290.Word-Pattern.py
# https://leetcode.com/problems/word-pattern/
#
# algorithms
# Easy (35.62%)
# Total Accepted:    155,351
# Total Submissions: 436,191


class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        arr = str.split(' ')
        if len(arr) != len(pattern):
            return False

        ch_2_word, word_2_ch = {}, {}

        for idx, ch in enumerate(pattern):
            if ch not in ch_2_word:
                ch_2_word[ch] = arr[idx]

            elif ch_2_word[ch] != arr[idx]:
                return False

            if arr[idx] not in word_2_ch:
                word_2_ch[arr[idx]] = ch

            elif word_2_ch[arr[idx]] != ch:
                return False

        return True
